File: src/main/java/com/letgo/book/domain/Book.java
package com.letgo.book.domain;

import com.letgo.shared.domain.AggregateRoot;

final public class Book extends AggregateRoot {
    private final BookId id;
    private BookTitle title;

    private Book(BookId id, BookTitle title) {
        this.id = id;
        this.title = title;
        storeEvent(new BookCreated(id.value(), title.value()));
    }

    public static Book create(String id, String title) {
        return new Book(BookId.create(id), BookTitle.create(title));
    }

    public static Book create(BookId id, BookTitle title) {
        return new Book(id, title);
    }

    public BookId id() {
        return id;
    }

    public BookTitle title() {
        return title;
    }

    public void changeTitle(BookTitle newTitle) {
        storeEvent(new BookTitleChanged(id.value(), title.value(), newTitle.value()));
        title = newTitle;
    }
}

File: src/main/kotlin/com/letgo/book/domain/Book.kt
package com.letgo.book.domain

import com.letgo.shared.domain.AggregateRoot

data class Book(
    private val id: BookId,
    private var status: BookStatus,
    private var title: BookTitle,
) : AggregateRoot() {
    constructor(id: String, title: String, createdAt: String) : this(BookId(id), BookTitle(title, createdAt))
    constructor(id: BookId, title: BookTitle) : this(id, BookStatus.Created, title)

    init {
        storeEvent(BookCreated(id.value(), title.value(), title.createdAt()))
    }

    fun id(): BookId = id

    fun title(): BookTitle = title

    fun hasBeenEdited(): Boolean = status == BookStatus.Edited

    fun canChangeTitle(newTitle: BookTitle): Boolean = title != newTitle && newTitle.isNewerThan(title)

    fun changeTitle(newTitle: BookTitle) {
        storeEvent(BookTitleChanged(id.value(), title.value(), newTitle.value(), newTitle.createdAt()))
        title = newTitle
        status = BookStatus.Edited
    }
}
